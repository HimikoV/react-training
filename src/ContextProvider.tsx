import { createContext, useEffect, useState, useMemo } from "react";
import { mapBasketData } from "./mappers/mapBasketData";
import { getDeliveryOptions } from "./service/getDeliveryOptions";
import { getOrderData } from "./service/getOrderData";
import { getPaymentOptions } from "./service/getPaymentOptions";
import { requestData } from "./service/requestData";
import { basketType } from "./types/basketType";
import { finalizeOptionsType } from "./types/finalizeOptionsType";
import { orderType } from "./types/orderType";
import { productType } from "./types/productType";
import { childrenInterface } from "./interfaces/childrenInterface";
import { basketProcessedType } from "./types/basketProcessedType";
import { contextProviderInterface } from "./interfaces/contextProviderInterface";

export const dataContext = createContext<contextProviderInterface>(
  {} as contextProviderInterface
);

const ContextProvider = ({ children }: childrenInterface) => {
  const [basketData, setBasketData] = useState<basketType[]>(
    JSON.parse(localStorage.getItem("basketData") || "[]")
  );
  const [filter, setFilter] = useState<string[]>([]);
  const [data, setData] = useState<productType[]>([]);
  const [loadLimit, setLoadLimit] = useState<number>(9);
  const [basketProcessedData, setBasketProcessedData] = useState<
    basketProcessedType[]
  >([]);
  const [deliveryOptions, setDeliveryOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [paymentOptions, setPaymentOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [open, setOpen] = useState<boolean>(false);
  const [orderData, setOrderData] = useState<orderType[]>([]);

  useEffect(() => {
    const fetchData = async () => {
      const responseData = await requestData();
      setData(responseData);
      const responseDelivery = await getDeliveryOptions();
      setDeliveryOptions(responseDelivery);
      const responsePayment = await getPaymentOptions();
      setPaymentOptions(responsePayment);
      const responseOrderData = await getOrderData();
      setOrderData(responseOrderData);
    };
    fetchData();
  }, []);

  useEffect(() => {
    localStorage.setItem("basketData", JSON.stringify(basketData));
  }, [basketData]);

  useEffect(() => {
    mapBasketData(data, basketData, setBasketProcessedData);
  }, [data]);

  const trueValues = useMemo(
    () => ({
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    }),
    [
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    ]
  );

  // const values = {};
  return (
    <dataContext.Provider value={trueValues}>{children}</dataContext.Provider>
  );
};
export default ContextProvider;
