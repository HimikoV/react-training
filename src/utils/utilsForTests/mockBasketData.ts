import { basketProcessedType } from "../../types/basketProcessedType";
import { basketType } from "../../types/basketType";

export const mockBasketData: basketType[] = [
  { id: "item_1", quantity: 4 },
  { id: "item_2", quantity: 5 },
  { id: "item_3", quantity: 1 },
];
export const mockBasketDataResult: basketType[] = [];
export const mockBasketProcessedData: basketProcessedType[] = [];
export const mockBasketProcessedDataResult: basketProcessedType[] = [
  {
    name: "product1",
    price: 15,
    category: "category1",
    id: "item_1",
    imgUrl: "url1",
    quantity: 4,
  },
  {
    name: "product2",
    price: 15,
    category: "category2",
    id: "item_2",
    imgUrl: "url2",
    quantity: 5,
  },
  {
    name: "product3",
    price: 15,
    category: "category3",
    id: "item_3",
    imgUrl: "url2",
    quantity: 1,
  },
];
