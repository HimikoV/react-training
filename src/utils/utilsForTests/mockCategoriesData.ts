import { categoryType } from "../../types/categoryType";

export const mockCategoriesData: categoryType[] = [
  {
    categoryName: "category1",
    categoryId: "category1",
  },
  {
    categoryName: "category2",
    categoryId: "category2",
  },
];
