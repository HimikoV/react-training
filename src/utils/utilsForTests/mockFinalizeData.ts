import { finalizeOptionsType } from "../../types/finalizeOptionsType";

export const mockFinalizeData: finalizeOptionsType[] = [
  {
    name: "DPD Courier",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: true,
  },
  {
    name: "DPD Inpost",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: false,
  },
];
export const mockFinalizeDataPayment: finalizeOptionsType[] = [
  {
    name: "Cash",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: true,
  },
  {
    name: "Card",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: false,
  },
];
export const mockFinalizeDataResult: finalizeOptionsType[] = [
  {
    name: "DPD Courier",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: true,
    key: "DPD Courier",
  },
  {
    name: "DPD Inpost",
    imgUrl:
      "https://images.unsplash.com/photo-1601647998802-5c86158216f6?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=755&q=80",
    selected: false,
    key: "DPD Inpost",
  },
];
