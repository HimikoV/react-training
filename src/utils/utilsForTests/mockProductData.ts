import { productType } from "../../types/productType";

export const mockProductData: productType[] = [
  {
    name: "product1",
    price: 15,
    category: "category1",
    id: "item1",
    imgUrl: "url1",
  },
  {
    name: "product2",
    price: 15,
    category: "category2",
    id: "item2",
    imgUrl: "url2",
  },
];
export const mockProductDataResult: productType[] = [
  {
    name: "product1",
    price: 15,
    category: "category1",
    id: "item1",
    imgUrl: "url1",
    key: "item1",
  },

  {
    name: "product2",
    price: 15,
    category: "category2",
    id: "item2",
    imgUrl: "url2",
    key: "item2",
  },
];
