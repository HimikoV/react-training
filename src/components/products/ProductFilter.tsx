import {
  Box,
  ToggleButton,
  IconButton,
  ToggleButtonGroup,
} from "@mui/material";
import FilterListIcon from "@mui/icons-material/FilterList";
import { useState, useEffect, useContext } from "react";
import SortByAlphaIcon from "@mui/icons-material/SortByAlpha";
import SortIcon from "@mui/icons-material/Sort";
import { requestCategories } from "../../service/requestCategories";
import { categoryType } from "../../types/categoryType";
import { dataContext } from "../../ContextProvider";

const ProductFilter = ({
  sortByLetter,
  sortByPrice,
  filterByCategory,
}: any) => {
  const [expand, setExpand] = useState<boolean>(false);
  const [categories, setCategories] = useState<categoryType[]>(
    [] as categoryType[]
  );
  const { filter } = useContext(dataContext);
  const fetchData = async () => {
    const responseData = await requestCategories();
    setCategories(responseData);
  };

  useEffect(() => {
    fetchData();
  }, []);
  return (
    <Box
      sx={{
        display: "flex",
        flexWrap: "wrap",
        p: 2,
        maxWidth: 500,
        justifyContent: "center",
        alignItems: "center",
        bgcolor: "#f7f4e1",
      }}
    >
      {categories.map((item: categoryType, index: number) => (
        <ToggleButtonGroup value={filter} key={item.categoryId}>
          <ToggleButton
            onClick={() => {
              filterByCategory(item.categoryId);
            }}
            value={item.categoryId}
            sx={{ m: 0.4, p: 0.6 }}
          >
            {item.categoryName}
          </ToggleButton>
        </ToggleButtonGroup>
      ))}

      <IconButton onClick={() => setExpand(!expand)}>
        <FilterListIcon />
      </IconButton>
      {expand ? (
        <>
          <IconButton onClick={() => sortByPrice()}>
            <SortIcon />
          </IconButton>
          <IconButton onClick={() => sortByLetter()}>
            <SortByAlphaIcon />
          </IconButton>
        </>
      ) : null}
    </Box>
  );
};

export default ProductFilter;
