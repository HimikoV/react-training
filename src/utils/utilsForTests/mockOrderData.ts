import { deliveryInformationType } from "../../types/deliveryInformationType";
import { orderType } from "../../types/orderType";
import { mockBasketData } from "./mockBasketData";

export const mockDeliveryInformation: deliveryInformationType = {
  name: "asdfasd",
  surname: "sdafas",
  email: "asdf@asdfjha.com",
  phone: "48325235235",
  zipcode: "23-325",
  state: "afdas",
  city: "asdfas",
};

export const mockOrderData: orderType[] = [
  {
    id: 1234,
    items: mockBasketData,
    date: "1/1/2022",
    delivery_information: mockDeliveryInformation,
    delivery_option: "DHL",
    payment_option: "cash",
  },
  {
    id: 1235,
    items: mockBasketData,
    date: "1/1/2022",
    delivery_information: mockDeliveryInformation,
    delivery_option: "DHL",
    payment_option: "cash",
  },
];
