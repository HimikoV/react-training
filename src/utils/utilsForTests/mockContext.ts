import { mockBasketData, mockBasketProcessedData } from "./mockBasketData";
import { mockProductData, mockProductDataResult } from "./mockProductData";
import { mockFinalizeData } from "./mockFinalizeData";
import { mockOrderDataArray } from "./mockOrderData";
import { contextProviderInterface } from "../../interfaces/contextProviderInterface";

export const mockContext: contextProviderInterface = {
  filter: [],
  setFilter: jest.fn(),
  basketData: mockBasketData,
  setBasketData: jest.fn(),
  data: mockProductData,
  loadLimit: 9,
  setLoadLimit: jest.fn(),
  basketProcessedData: mockBasketProcessedData,
  setBasketProcessedData: jest.fn(),
  deliveryOptions: mockFinalizeData,
  setDeliveryOptions: jest.fn(),
  paymentOptions: mockFinalizeData,
  setPaymentOptions: jest.fn(),
  open: false,
  setOpen: jest.fn(),
  orderData: mockOrderDataArray,
  setOrderData: jest.fn(),
  order: false,
  setOrder: jest.fn(),
  sort: "letter",
  setSort: jest.fn(),
  processedData: mockProductDataResult,
  setProcessedData: jest.fn(),
  orderLoadLimit: 20,
  setOrderLoadLimit: jest.fn(),
};
