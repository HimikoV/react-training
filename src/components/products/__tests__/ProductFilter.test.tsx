import ProductFilter from "../ProductFilter";
import { describe, it } from "@jest/globals";
import renderer from "react-test-renderer";
import { dataContext } from "../../../ContextProvider";
import { mockContext } from "../../../utils/utilsForTests/mockContext";

describe("testing ProductFilter render", () => {
  it("ProductFilter Renders Correctly", () => {
    const productFilter = renderer
      .create(
        <dataContext.Provider value={mockContext}>
          <ProductFilter filterByCategory={(arg: string) => {}} />
        </dataContext.Provider>
      )
      .toJSON();
    expect(productFilter).toMatchSnapshot();
  });
});
