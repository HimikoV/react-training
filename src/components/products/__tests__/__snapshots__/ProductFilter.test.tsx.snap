// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ProductFilter render ProductFilter Renders Correctly 1`] = `
<div
  className="MuiGrid-root css-er906r-MuiGrid-root"
>
  <button
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="FilterListIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
      />
    </svg>
  </button>
</div>
`;
